# <div style="text-align:center">2023년 봄학기 파이썬데이터분석 - 2주차 <br/><pandas 입문></div>

<div style="text-align: right"> <b>강남대학교 인공지능융합공학부 곽찬희</b></div>
<div style="text-align: right"> <b>chk at kangnam dot ac dot kr</b></div>

# 0. 개요

1주차에는 pandas 의 기본적인 사항들에 대해 알아보고, 익숙해지는 시간을 갖겠습니다.

시작 전에 ecampus 의 자료로 올라가 있는, data.zip 파일을 압축해제하기 바랍니다.


# 1. Pandas 소개

pandas 는 python 의 대표적인 데이터 분석용 라이브러리입니다. 이를 활용하여 여러분들은 데이터를 구조화하고 조작할 수 있습니다.

먼저 pandas 를 불러와 보겠습니다.

보통 numpy 와 함께 로드하며, 각각 pd와 np 라는 별명을 주로 사용합니다.



In [8]:
import pandas as pd
import numpy as np

## 1.1. pandas DataFrame

pandas 는 DataFrame 과 Series 의 데이터 구조를 갖고 있습니다.

DataFrame (DF) 은 표 형태의 2차원 데이터 구조로, 행(인덱스), 열, 데이터 를 갖습니다.
Series 는 1차원의 데이터 구조로 인덱스와 데이터만 갖습니다.

그러니까, 다음과 같이 DF가 존재할 때, 각 열은 Series 로 생각해볼 수 있겠죠.

![Image](https://bookdata.readthedocs.io/en/latest/_images/base_01_pandas_5_0.png)

<https://bookdata.readthedocs.io/en/latest/_images/base_01_pandas_5_0.pnghttps://bookdata.readthedocs.io/en/latest/_images/base_01_pandas_5_0.png>

## 1.2. DF 구경하기

그러면 데이터를 불러와서 DF를 만들어보겠습니다.

아래의 코드를 볼까요?

In [9]:
movies = pd.read_csv('movie.csv')
type(movies)

pandas.core.frame.DataFrame

In [10]:
movies

color      director_name  num_critic_for_reviews  duration  \
0     Color      James Cameron                   723.0     178.0   
1     Color     Gore Verbinski                   302.0     169.0   
2     Color         Sam Mendes                   602.0     148.0   
3     Color  Christopher Nolan                   813.0     164.0   
4       NaN        Doug Walker                     NaN       NaN   
...     ...                ...                     ...       ...   
4911  Color        Scott Smith                     1.0      87.0   
4912  Color                NaN                    43.0      43.0   
4913  Color   Benjamin Roberds                    13.0      76.0   
4914  Color        Daniel Hsia                    14.0     100.0   
4915  Color           Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                         0.0                   855.0  Joel David Moore   
1                       563.0                  1000.0     Orlando Bloom   
2                         0.0                   161.0      Rory Kinnear   
3                     22000.0                 23000.0    Christian Bale   
4                       131.0                     NaN        Rob Walker   
...                       ...                     ...               ...   
4911                      2.0                   318.0     Daphne Zuniga   
4912                      NaN                   319.0     Valorie Curry   
4913                      0.0                     0.0     Maxwell Moody   
4914                      0.0                   489.0     Daniel Henney   
4915                     16.0                    16.0  Brian Herzlinger   

      actor_1_facebook_likes        gross                           genres  \
0                     1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                    40000.0  309404152.0         Action|Adventure|Fantasy   
2                    11000.0  200074175.0        Action|Adventure|Thriller   
3                    27000.0  448130642.0                  Action|Thriller   
4                      131.0          NaN                      Documentary   
...                      ...          ...                              ...   
4911                   637.0          NaN                     Comedy|Drama   
4912                   841.0          NaN     Crime|Drama|Mystery|Thriller   
4913                     0.0          NaN            Drama|Horror|Thriller   
4914                   946.0      10443.0             Comedy|Drama|Romance   
4915                    86.0      85222.0                      Documentary   

      ... num_user_for_reviews language  country  content_rating       budget  \
0     ...               3054.0  English      USA           PG-13  237000000.0   
1     ...               1238.0  English      USA           PG-13  300000000.0   
2     ...                994.0  English       UK           PG-13  245000000.0   
3     ...               2701.0  English      USA           PG-13  250000000.0   
4     ...                  NaN      NaN      NaN             NaN          NaN   
...   ...                  ...      ...      ...             ...          ...   
4911  ...                  6.0  English   Canada             NaN          NaN   
4912  ...                359.0  English      USA           TV-14          NaN   
4913  ...                  3.0  English      USA             NaN       1400.0   
4914  ...                  9.0  English      USA           PG-13          NaN   
4915  ...                 84.0  English      USA              PG       1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0         2009.0                  936.0        7.9          1.78   
1         2007.0                 5000.0        7.1          2.35   
2         2015.0                  393.0        6.8          2.35   
3         2012.0                23000.0        8.5          2.35   
4            NaN                   12.

movies 에 read_csv 함수의 결과가 들어갔습니다. 
csv 는 comma seperated values 의 약자로, 쉼표로 구분한 데이터의 형태를 의미합니다.
쉽게 생각하면, 엑셀의 내용만 저장한 형태가 되는 것이죠.

- type 로 확인해보면 DataFrame인 것을 알 수 있죠? 그러니까, read_csv 의 결과는 DF가 된다 를 기억합시다.

- NaN은 Not a Number 의 약자인데, 여기서는 결측치 (missing value)를 의미합니다. 나중에 어떻게 처리하는지 배울겁니다.

- ... 은 데이터가 생략되어 있음을 의미합니다.

- Python 의 규칙에 따라, 숫자는 0부터 세는 것을 기억하세요.

보통 데이터 전체를 출력하는 경우는 많지 않고, 일부분만 봅니다. (화면에 다 안 나오니까요)

head 나 tail 함수를 사용해볼 수 있는데 용법은 다음과 같습니다.

>movie.head() #상위 5줄

>movie.head(10) # 상위 10줄

한번 해볼까요?

In [11]:
movies.head(7)

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   
5  Color     Andrew Stanton                   462.0     132.0   
6  Color          Sam Raimi                   392.0     156.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   
5                    475.0                   530.0   Samantha Morton   
6                      0.0                  4000.0      James Franco   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   
5                   640.0   73058679.0          Action|Adventure|Sci-Fi  ...   
6                 24000.0  336530303.0         Action|Adventure|Romance  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   
5                738.0  English      USA           PG-13  263700000.0   
6               1902.0  English      USA           PG-13  258000000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   
5      2012.0                  632.0        6.6          2.35   
6      2007.0                11000.0        6.2          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  
5                24000  
6                    0  

[7 rows x 28 columns]

In [12]:
movies.tail(10)

color     director_name  num_critic_for_reviews  duration  \
4906  Color     Shane Carruth                   143.0      77.0   
4907  Color  Neill Dela Llana                    35.0      80.0   
4908  Color  Robert Rodriguez                    56.0      81.0   
4909  Color   Anthony Vallone                     NaN      84.0   
4910  Color      Edward Burns                    14.0      95.0   
4911  Color       Scott Smith                     1.0      87.0   
4912  Color               NaN                    43.0      43.0   
4913  Color  Benjamin Roberds                    13.0      76.0   
4914  Color       Daniel Hsia                    14.0     100.0   
4915  Color          Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes        actor_2_name  \
4906                    291.0                     8.0      David Sullivan   
4907                      0.0                     0.0     Edgar Tancangco   
4908                      0.0                     6.0     Peter Marquardt   
4909                      2.0                     2.0      John Considine   
4910                      0.0                   133.0  Caitlin FitzGerald   
4911                      2.0                   318.0       Daphne Zuniga   
4912                      NaN                   319.0       Valorie Curry   
4913                      0.0                     0.0       Maxwell Moody   
4914                      0.0                   489.0       Daniel Henney   
4915                     16.0                    16.0    Brian Herzlinger   

      actor_1_facebook_likes      gross                               genres  \
4906                   291.0   424760.0                Drama|Sci-Fi|Thriller   
4907                     0.0    70071.0                             Thriller   
4908                   121.0  2040920.0  Action|Crime|Drama|Romance|Thriller   
4909                    45.0        NaN                          Crime|Drama   
4910                   296.0     4584.0                         Comedy|Drama   
4911                   637.0        NaN                         Comedy|Drama   
4912                   841.0        NaN         Crime|Drama|Mystery|Thriller   
4913                     0.0        NaN                Drama|Horror|Thriller   
4914                   946.0    10443.0                 Comedy|Drama|Romance   
4915                    86.0    85222.0                          Documentary   

      ... num_user_for_reviews language      country  content_rating  budget  \
4906  ...                371.0  English          USA           PG-13  7000.0   
4907  ...                 35.0  English  Philippines       Not Rated  7000.0   
4908  ...                130.0  Spanish          USA               R  7000.0   
4909  ...                  1.0  English          USA           PG-13  3250.0   
4910  ...                 14.0  English          USA       Not Rated  9000.0   
4911  ...                  6.0  English       Canada             NaN     NaN   
4912  ...                359.0  English          USA           TV-14     NaN   
4913  ...                  3.0  English          USA             NaN  1400.0   
4914  ...                  9.0  English          USA           PG-13     NaN   
4915  ...                 84.0  English          USA              PG  1100.0   

      title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
4906      2004.0                   45.0        7.0          1.85   
4907      2005.0                    0.0        6.3           NaN   
4908      1992.0                   20.0        6.9          1.37   
4909      2005.0                   44.0        7.8           NaN   
4910      2011.0                  205.0        6.4           NaN   
4911      2013.0                  470.0        7.7           NaN   
4912         NaN                  593.0        7.5         16.00   
4913      2013.0                    0.0        6.3           NaN   
4914      2012.0                  719.0       

앞서 DF는 세 요쇼로 구성된다고 했던 것 기억 나시나요?

열, 인덱스(행), 그리고 데이터 였죠.

이를 각각 접근할 수 있습니다.

아래의 코드로 확인하세요.

In [13]:
columns = movies.columns
#list(columns)
pd.Series(columns)

0                         color
1                 director_name
2        num_critic_for_reviews
3                      duration
4       director_facebook_likes
5        actor_3_facebook_likes
6                  actor_2_name
7        actor_1_facebook_likes
8                         gross
9                        genres
10                 actor_1_name
11                  movie_title
12              num_voted_users
13    cast_total_facebook_likes
14                 actor_3_name
15         facenumber_in_poster
16                plot_keywords
17              movie_imdb_link
18         num_user_for_reviews
19                     language
20                      country
21               content_rating
22                       budget
23                   title_year
24       actor_2_facebook_likes
25                   imdb_score
26                 aspect_ratio
27         movie_facebook_likes
dtype: object

DF.columns 를 하면 해당 DF 가 갖고 있는 열의 이름이 추출됩니다. 데이터 형식이 Index 로 나오는 것을 기억하세요. 

인덱스도 같이 볼까요?

In [14]:
index = movies.index
index

RangeIndex(start=0, stop=4916, step=1)

인덱스는 row 의 번호라고 생각할 수 있습니다. 

아까 확인했듯, 0번부터 4916번까지가 인덱스로 들어갑니다. 

위와는 다르게, RangeIndex 로 구성이 되어 있지요? 이는 연속된 숫자를 인덱스로 활용하기 때문입니다.

데이터도 마지막으로 살펴보겠습니다.

여기서는 numpy형태의 ndarray 로 나옵니다.

In [15]:
data = movies.to_numpy()
data

array([['Color', 'James Cameron', 723.0, ..., 7.9, 1.78, 33000],
       ['Color', 'Gore Verbinski', 302.0, ..., 7.1, 2.35, 0],
       ['Color', 'Sam Mendes', 602.0, ..., 6.8, 2.35, 85000],
       ...,
       ['Color', 'Benjamin Roberds', 13.0, ..., 6.3, nan, 16],
       ['Color', 'Daniel Hsia', 14.0, ..., 6.3, 2.35, 660],
       ['Color', 'Jon Gunn', 43.0, ..., 6.6, 1.85, 456]], dtype=object)

## 1.3. 데이터 형식 

pandas 의 데이터 형식은 다음과 같은 종류가 있습니다.
각 컬럼이 어떤 형식인지 정확하게 아는 것이 중요하니, 눈여겨보세요.
결측치를 표시할 수 있는지 여부에 따라 달라집니다.

* float : 부동소수점 (결측치 O) 
* int : 정수 (결측치 X)
* Int64 : 정수 (결측치 O)
* object : 문자열
* category : 범주형 (결측치 O)
* bool : 불리언 (결측치 X)
* boolean : 불리언 (결측치 O)
* datetime64 : 날짜 (결측치 O)

movies 의 컬럼들은 어떤 타입인지 볼까요?

DF.dtypes 로 확인할 수 있습니다.

In [16]:
movies.dtypes

color                         object
director_name                 object
num_critic_for_reviews       float64
duration                     float64
director_facebook_likes      float64
actor_3_facebook_likes       float64
actor_2_name                  object
actor_1_facebook_likes       float64
gross                        float64
genres                        object
actor_1_name                  object
movie_title                   object
num_voted_users                int64
cast_total_facebook_likes      int64
actor_3_name                  object
facenumber_in_poster         float64
plot_keywords                 object
movie_imdb_link               object
num_user_for_reviews         float64
language                      object
country                       object
content_rating                object
budget                       float64
title_year                   float64
actor_2_facebook_likes       float64
imdb_score                   float64
aspect_ratio                 float64
m

dtypes 의 결과는 Series 이기 때문에, 값을 셀 수 있는 `value_counts` 함수를 사용할 수 있습니다. 아래를 볼까요?

In [17]:
movies.dtypes.value_counts()

float64    13
object     12
int64       3
dtype: int64

`info` 함수를 사용해도 비슷한 결과를 얻습니다.

In [18]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4916 entries, 0 to 4915
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      4897 non-null   object 
 1   director_name              4814 non-null   object 
 2   num_critic_for_reviews     4867 non-null   float64
 3   duration                   4901 non-null   float64
 4   director_facebook_likes    4814 non-null   float64
 5   actor_3_facebook_likes     4893 non-null   float64
 6   actor_2_name               4903 non-null   object 
 7   actor_1_facebook_likes     4909 non-null   float64
 8   gross                      4054 non-null   float64
 9   genres                     4916 non-null   object 
 10  actor_1_name               4909 non-null   object 
 11  movie_title                4916 non-null   object 
 12  num_voted_users            4916 non-null   int64  
 13  cast_total_facebook_likes  4916 non-null   int64

## 1.4. 열과 행의 선택

DF는 표 형태이기 때문에 열 혹은 행을 기준으로 데이터를 선택할 수 있습니다.

기본적으로 numpy 의 array 를 선택하는 것과 동일하게 데이터를 있는데요.

다음과 같은 방법이 있습니다.

'A'라는 열이 첫 열로 df에 있을 때,

> df.A

> df['A']

> df.loc[:, 'A']

> df.iloc[:, 0]

1번과 2번 방법은 직관적이지요?

3번 loc은 레이블 이름 기반으로 데이터에 접근합니다.
이름표를 보고 가는 것이지요.

4번 iloc은 위치 기반으로 데이터에 접근합니다.
데이터의 위치번호를 보고 접근합니다.

그리고, 두 개의 차원을 표시할 때 :는 모든 것을 의미합니다.
즉, 모든 행을 선택하는데 열 이름이 'A'인 것 이 되겠죠.

movies 의 director_name 컬럼에 접근해볼까요?

In [19]:
movies.director_name

0           James Cameron
1          Gore Verbinski
2              Sam Mendes
3       Christopher Nolan
4             Doug Walker
              ...        
4911          Scott Smith
4912                  NaN
4913     Benjamin Roberds
4914          Daniel Hsia
4915             Jon Gunn
Name: director_name, Length: 4916, dtype: object

In [20]:
movies.loc[:, 'director_name']

0           James Cameron
1          Gore Verbinski
2              Sam Mendes
3       Christopher Nolan
4             Doug Walker
              ...        
4911          Scott Smith
4912                  NaN
4913     Benjamin Roberds
4914          Daniel Hsia
4915             Jon Gunn
Name: director_name, Length: 4916, dtype: object

In [63]:
movies.iloc[:, 1] #iloc 함수 뒤에는 숫자를 포함하지 않음

0           James Cameron
1          Gore Verbinski
2              Sam Mendes
3       Christopher Nolan
4             Doug Walker
              ...        
4911          Scott Smith
4912                  NaN
4913     Benjamin Roberds
4914          Daniel Hsia
4915             Jon Gunn
Name: director_name, Length: 4916, dtype: object

이렇게 출력되는 열은 모두 pandas Series 형태를 갖습니다.

type 으로 확인해볼까요?

In [22]:
type(movies.director_name)

pandas.core.series.Series

여러 방법이 있지만 가장 타이핑이 적은(덜 귀찮은) 방법은 .을 쓰는 것이죠.

게다가 jupyterlab 에서는 자동완성도 지원하니 참 편합니다.

다만, 일부 특수어를 컬럼 이름으로 가진 경우나, 컬럼 이름에 빈칸이 있는 경우엔 반드시 df['A'] 를 사용해야 합니다.

또한 새로운 컬럼을 만들 때에도 . 은 사용할 수 없습니다.

보통 간단한 Series 를 추출할 때엔 . 를 사용합니다.



## 1.5. Series 와 함수들

Series 도 마찬가지로 너무 긴 경우 다 표현할 필요는 없죠.

head / tail 을 사용할 수 있지만, 여기서는 sample 이란 함수도 배워볼게요.

말 그대로 sampling 을 해주는 (몇 개만 추출해주는) 함수입니다.

> s.sample(n=?, random_sate=?)

n은 샘플링 할 개수를 의미하고, random_state 는 random 을 컨트롤하기 위해 넣는 숫자입니다.



In [23]:
director = movies.director_name
fb_likes = movies.actor_1_facebook_likes

In [24]:
director.sample(3, random_state=42)

2347    Brian Percival
4687       Lucio Fulci
691      Phillip Noyce
Name: director_name, dtype: object

---
### - value_counts - 값의 개수 세기

value_counts 는 가장 자주 쓸 함수 중 하나입니다.

말 그대로 값의 개수를 세어줍니다.

일반적으로는 명목형/factor 컬럼에 적용합니다.

director 에 적용시켜보세요.

In [25]:
director.head()

0        James Cameron
1       Gore Verbinski
2           Sam Mendes
3    Christopher Nolan
4          Doug Walker
Name: director_name, dtype: object

In [26]:
director.value_counts()

Steven Spielberg    26
Woody Allen         22
Martin Scorsese     20
Clint Eastwood      20
Ridley Scott        16
                    ..
John Putch           1
Luca Guadagnino      1
Sam Fell             1
Dan Fogelman         1
Daniel Hsia          1
Name: director_name, Length: 2397, dtype: int64

normalize=True 의 인자를 넘겨주게 되면, 비율을 구해줍니다.

같은 결과를 보죠.


In [27]:
director.value_counts(normalize=True)

Steven Spielberg    0.005401
Woody Allen         0.004570
Martin Scorsese     0.004155
Clint Eastwood      0.004155
Ridley Scott        0.003324
                      ...   
John Putch          0.000208
Luca Guadagnino     0.000208
Sam Fell            0.000208
Dan Fogelman        0.000208
Daniel Hsia         0.000208
Name: director_name, Length: 2397, dtype: float64

### - describe : 기술 통계

describe 를 사용하면 수치형 데이터의 기술 통계를 구할 수 있습니다.

df 와 s 모두 적용 가능합니다.

명목형 혹은 string 컬럼에 대해서는 count 와 unique , top 값이 나옵니다.


In [28]:
movies.describe() #head -> tail -> describe 순서로 데이터 확인

num_critic_for_reviews     duration  director_facebook_likes  \
count             4867.000000  4901.000000              4814.000000   
mean               137.988905   107.090798               691.014541   
std                120.239379    25.286015              2832.954125   
min                  1.000000     7.000000                 0.000000   
25%                 49.000000    93.000000                 7.000000   
50%                108.000000   103.000000                48.000000   
75%                191.000000   118.000000               189.750000   
max                813.000000   511.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             4893.000000             4909.000000  4.054000e+03   
mean               631.276313             6494.488491  4.764451e+07   
std               1625.874802            15106.986884  6.737255e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                132.000000              607.000000  5.019656e+06   
50%                366.000000              982.000000  2.504396e+07   
75%                633.000000            11000.000000  6.110841e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     4.916000e+03                4916.000000           4903.000000   
mean      8.264492e+04                9579.815907              1.377320   
std       1.383222e+05               18164.316990              2.023826   
min       5.000000e+00                   0.000000              0.000000   
25%       8.361750e+03                1394.750000              0.000000   
50%       3.313250e+04                3049.000000              1.000000   
75%       9.377275e+04               13616.750000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           4895.000000  4.432000e+03  4810.000000   
mean             267.668846  3.654749e+07  2002.447609   
std              372.934839  1.002427e+08    12.453977   
min                1.000000  2.180000e+02  1916.000000   
25%               64.000000  6.000000e+06  1999.000000   
50%              153.000000  1.985000e+07  2005.000000   
75%              320.500000  4.300000e+07  2011.000000   
max             5060.000000  4.200000e+09  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             4903.000000  4916.000000   4590.000000           4916.000000  
mean              1621.923516     6.437429      2.222349           7348.294142  
std               4011.299523     1.127802      1.402940          19206.016458  
min                  0.000000     1.600000      1.180000              0.000000  
25%                277.000000     5.800000      1.850000              0.000000  
50%                593.000000     6.600000      2.350000            159.000000  
75%                912.000000     7.200000      2.350000           2000.000000  
max             137000.000000     9.500000     16.000000         349000.000000

In [29]:
director.describe()

count                 4814
unique                2397
top       Steven Spielberg
freq                    26
Name: director_name, dtype: object


여기서 row index 로 등장하는 것의 대부분은 함수로 추출할 수 있습니다.

예를 들어 count, mean, std, min, max, unique 와 같은 것들은 함수로 해당하는 값을 구할 수 있죠.

몇 가지만 코드로 구현해봅시다.

In [30]:
fb1 = movies.actor_1_facebook_likes

print(fb1.min(), fb1.max(), fb1.mean())

0.0 640000.0 6494.488490527602


### NA 값 확인하기 isna / fillna / dropna

info 에서 null 이 아닌 값의 개수를 확인했었죠?

null의 수를 세는 또 다른 방법은 isna 함수를 사용하는 겁니다. 

isna 는 na 인 경우 True를, 그렇지 않은 값은 False 를 반환합니다.

아래와 같이 말이죠.



In [31]:
movies.isna().sum() #isna 함수는 n/a 값을 T or F로 표시

color                         19
director_name                102
num_critic_for_reviews        49
duration                      15
director_facebook_likes      102
actor_3_facebook_likes        23
actor_2_name                  13
actor_1_facebook_likes         7
gross                        862
genres                         0
actor_1_name                   7
movie_title                    0
num_voted_users                0
cast_total_facebook_likes      0
actor_3_name                  23
facenumber_in_poster          13
plot_keywords                152
movie_imdb_link                0
num_user_for_reviews          21
language                      12
country                        5
content_rating               300
budget                       484
title_year                   106
actor_2_facebook_likes        13
imdb_score                     0
aspect_ratio                 326
movie_facebook_likes           0
dtype: int64

In [32]:
test_series = pd.Series([1, 3, 6, np.nan, 1])
test_series.isna()

0    False
1    False
2    False
3     True
4    False
dtype: bool

결측치를 채울 때엔 fillna 함수를 사용합니다.

위의 test_series 의 na 값을 채워볼게요.

> series.fillna(채울값)

의 형태로 사용합니다.

In [33]:
test_series.fillna(0)

0    1.0
1    3.0
2    6.0
3    0.0
4    1.0
dtype: float64

만약에 값을 대체하는 것이 아니라, 결측치를 삭제하고 싶을 땐 dropna 를 사용합니다.

아래를 참고하세요.


In [34]:
test_series.dropna()

0    1.0
1    3.0
2    6.0
4    1.0
dtype: float64

In [35]:
test_series = test_series.dropna()

In [36]:
test_series

0    1.0
1    3.0
2    6.0
4    1.0
dtype: float64

주의할 점은, 저렇게 바꿨다고 해서 test_series 가 바뀌지 않는다는 것 입니다.

test_series를 바꾸기 위해서는
> test_series = test_series.dropna() 

를 해주거나, 

> test_series.dropna(inplace=True)

로 바꿔줘야 합니다.

다만, 원본 데이터를 바꾼다는 것은 되돌릴 수 없는 일이 되는 것이니 신중하게 접근하세요.

## 1.6. Series 의 연산

Series 를 계산에 사용할 때엔 재미있는 특징이 있습니다.

바로 broadcasting 인데요.

방송을 송출하는 듯한 이 것은 무엇인지 예를 통해 봅시다.

In [37]:
a = pd.Series([1, 2, 3])

a + 1

0    2
1    3
2    4
dtype: int64

분명 a 는 1, 2, 3이 들어있는 Series 이기에,  (1, 2, 3) + 1 이라는 이상한 구조처럼 보이지만, 

계산되는 대상이 scalar 인 경우, 이를 자동적으로 Series 에 맞춰줍니다.

그러니까 실제로는, (1, 2, 3) + (1, 1, 1) 이 되어서 결과인 (2, 3, 4) 가 나온 것이죠.

이런 계산은 사칙연산 ( +, -, *, /, //, %) 뿐만 이나리 논리 비교( <, >, >=, <=, !) 에서도 동일하게 작용합니다.

하나 더 만들어볼게요.

In [38]:
a = pd.Series(np.arange(0, 10))
# a = [0, 1, 2, ..., 8, 9] % [2, 2, 2, 2, ... 2]
print(a % 2 == 0)

0     True
1    False
2     True
3    False
4     True
5    False
6     True
7    False
8     True
9    False
dtype: bool


위의 코드는 0부터 9까지의 수 중 2로 나눈 나머지가 0인, 즉 짝수인 경우를 확인하는 코드입니다.

a 는 전체 Series 를 의미하지만, 이를 식으로 표현할 때엔 마치 하나의 수인 것처럼 다룰 수 있죠.

이런 연산을 vector 화된 연산 (vectorization )이라고 하는데, for 문을 사용하는 것보다 훨씬 효율적인 접근입니다.

나중에 또 다룰 기회가 있을 겁니다.

---

동일한 계산을 함수를 사용해서도 가능합니다.

예를 들어 다음과 같죠.


In [39]:
fb_likes.add(1) # fb_likes + 1
fb_likes.sub(2) # fb_likes - 2

0         998.0
1       39998.0
2       10998.0
3       26998.0
4         129.0
         ...   
4911      635.0
4912      839.0
4913       -2.0
4914      944.0
4915       84.0
Name: actor_1_facebook_likes, Length: 4916, dtype: float64

첫 번째는 fb_likes + 1, 두 번째는 fb_likes - 2와 같습니다.

그러면 둘 사이엔 어떤 차이가 있는 것일까요?

함수를 이용해서 계산할 때엔 fill_value 라는 인자를 넘겨줄 수 있는데, 이를 통해 na 값을 처리할 수 있습니다.

na 값은 어떤 사칙연산을 적용하더라도 na 가 나옵니다. 아래의 결과를 보죠.


In [40]:
np.nan - 1

nan

그런데 fill_value 를 이용하면, fill_na 의 효과가 나타나면서 계산이 가능합니다.

다음과 같이 말이죠.


In [41]:
fb_likes.sub(2, fill_value=0)[fb_likes.isna()]

4401   -2.0
4418   -2.0
4608   -2.0
4721   -2.0
4822   -2.0
4823   -2.0
4864   -2.0
Name: actor_1_facebook_likes, dtype: float64

위의 내용은 na 값을 가진 것에 0을 채워 넣은 뒤 sub(2), 즉 2를 뺀 것이 됩니다.

여기서 기억할 두 가지

1. na 는 계산 불가능하다.
2. vector 연산에서 na 가 무시되는 것을 막기 위해서는 fillna 를 이용해 값을 먼저 채우거나, 계산 함수에서 fill_value 를 써준다.

## 1.7. 메서드 체인

위에서도 잠깐씩 등장했지만, pandas 에서는 메서드의 결과에 다시 메서드를 적용할 수 있습니다.

 다음과 같이 말이죠.

In [42]:
director

0           James Cameron
1          Gore Verbinski
2              Sam Mendes
3       Christopher Nolan
4             Doug Walker
              ...        
4911          Scott Smith
4912                  NaN
4913     Benjamin Roberds
4914          Daniel Hsia
4915             Jon Gunn
Name: director_name, Length: 4916, dtype: object

In [43]:
director.value_counts().head().tail()

Steven Spielberg    26
Woody Allen         22
Martin Scorsese     20
Clint Eastwood      20
Ridley Scott        16
Name: director_name, dtype: int64

director 는 Series 이고, Series 에 적용할 수 있는 함수 value_counts 를 적용했는데 이 결과 역시 Series 입니다.

따라서 여기에 head를 연속해 적용할 수 있게 되는 것이지요.

## 1.8. DF의 열 다루기 (1) - 열 이름 변경

열 이름을 변경할 때엔 rename 함수를 사용합니다.

다음의 예를 볼까요?



In [44]:
movies.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes'],
      dtype='object')

In [45]:
col_map = {'director_name':'director',
          'num_critic_for_reviews': 'critic_reviews'}

movies.rename(columns=col_map).iloc[:, :3].head()

color           director  critic_reviews
0  Color      James Cameron           723.0
1  Color     Gore Verbinski           302.0
2  Color         Sam Mendes           602.0
3  Color  Christopher Nolan           813.0
4    NaN        Doug Walker             NaN

col_map 이란 딕셔너리를 만들어서 기존 컬럼명:새 컬럼명 의 쌍을 만들어준 뒤, rename 함수에 적용합니다.

index(row 의 이름)에도 동일하게 적용할 수 있는데, 이 때는 rename 에 index=바꿀 이름들 을 넣어줍니다.

해당 자리에는 딕셔너리 말고 함수를 넣어줄 수도 있습니다.

다음의 구조를 볼까요?

In [46]:
movies.rename(columns=lambda x: x.strip().upper()).columns

Index(['COLOR', 'DIRECTOR_NAME', 'NUM_CRITIC_FOR_REVIEWS', 'DURATION',
       'DIRECTOR_FACEBOOK_LIKES', 'ACTOR_3_FACEBOOK_LIKES', 'ACTOR_2_NAME',
       'ACTOR_1_FACEBOOK_LIKES', 'GROSS', 'GENRES', 'ACTOR_1_NAME',
       'MOVIE_TITLE', 'NUM_VOTED_USERS', 'CAST_TOTAL_FACEBOOK_LIKES',
       'ACTOR_3_NAME', 'FACENUMBER_IN_POSTER', 'PLOT_KEYWORDS',
       'MOVIE_IMDB_LINK', 'NUM_USER_FOR_REVIEWS', 'LANGUAGE', 'COUNTRY',
       'CONTENT_RATING', 'BUDGET', 'TITLE_YEAR', 'ACTOR_2_FACEBOOK_LIKES',
       'IMDB_SCORE', 'ASPECT_RATIO', 'MOVIE_FACEBOOK_LIKES'],
      dtype='object')

lambda 함수를 이용해서 컬럼명을 모두 대문자로 바꿔보았습니다.

더 복잡한 함수를 적용해야 한다면, 새로 함수를 정의하고 그 함수를 사용할 수도 있겠죠!



## 1.9. DF의 열 다루기 (2) - 열 생성, 삭제

열을 만들 때엔, 이전에 배운 방식을 사용할 수 있습니다.

바로 df['열이름'] 이죠.

예를 들어, movies 에 새로운 열 A를 만든다고 합시다.

그리고 초기값을 모두 0으로 설정합니다.

In [47]:
movies['A'] = 0 #생성시 대괄호
movies.A 

0       0
1       0
2       0
3       0
4       0
       ..
4911    0
4912    0
4913    0
4914    0
4915    0
Name: A, Length: 4916, dtype: int64

In [48]:
movies.B = 0

In [49]:
movies.columns

Index(['color', 'director_name', 'num_critic_for_reviews', 'duration',
       'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name',
       'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name',
       'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
       'actor_3_name', 'facenumber_in_poster', 'plot_keywords',
       'movie_imdb_link', 'num_user_for_reviews', 'language', 'country',
       'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes',
       'imdb_score', 'aspect_ratio', 'movie_facebook_likes', 'A'],
      dtype='object')

기존 열들을 계산해 새로운 열을 만들어볼까요?

movies 데이터에 facebook 좋아요 수를 합쳐보겠습니다.
열 이름은 다음과 같습니다.

actor_1_facebook_likes
actor_2_facebook_likes
actor_3_facebook_likes
director_facebook_likes

먼저 이들을 합쳐서 하나의 Series 로 만든 뒤, 새 열로 넣어보겠습니다.

In [50]:
movies['total_likes'] = (movies.actor_1_facebook_likes + 
                         movies.actor_2_facebook_likes + 
                         movies.actor_3_facebook_likes + 
                         movies.director_facebook_likes)

In [51]:
movies.total_likes

0        2791.0
1       46563.0
2       11554.0
3       95000.0
4           NaN
         ...   
4911     1427.0
4912        NaN
4913        0.0
4914     2154.0
4915      141.0
Name: total_likes, Length: 4916, dtype: float64

동일한 결과를 sum 함수를 이용해 낼 수도 있습니다.

아래를 볼까요?

여기서 axis=1 은 컬럼 방향으로의 합을 의미합니다. 0일 때와 비교해보세요.

In [52]:
movies.isna().sum(axis=1) #n/a 값을 계산하고 싶을 때 sum 함수 이용

0        0
1        0
2        0
3        0
4       15
        ..
4911     4
4912     6
4913     4
4914     2
4915     0
Length: 4916, dtype: int64

In [53]:
cols = ['actor_1_facebook_likes',
        'actor_2_facebook_likes',
        'actor_3_facebook_likes',
        'director_facebook_likes']

sum_col = movies.loc[:,cols].sum(axis=1)
sum_col

0        2791.0
1       46563.0
2       11554.0
3       95000.0
4         274.0
         ...   
4911     1427.0
4912     1753.0
4913        0.0
4914     2154.0
4915      141.0
Length: 4916, dtype: float64

movies df 는 상당히 크기 때문에, facebook 이 컬럼에 들어가 있는 것만 선택해서 새로운 df를 만들어봅시다.

아래의 코드를 볼까요?

In [65]:
for c in movies.columns:
    #print(c)
    if 'facebook' in c:
        print(c) #아래함수와 같은 의미이나 지금껄 축약한게 아래꺼

director_facebook_likes
actor_3_facebook_likes
actor_1_facebook_likes
cast_total_facebook_likes
actor_2_facebook_likes
movie_facebook_likes


In [54]:
[
    c 
    for c in movies.columns 
    if 'facebook' in c
]

['director_facebook_likes',
 'actor_3_facebook_likes',
 'actor_1_facebook_likes',
 'cast_total_facebook_likes',
 'actor_2_facebook_likes',
 'movie_facebook_likes']

In [55]:
movies_fb = movies[[
    c 
    for c in movies.columns 
    if 'facebook' in c
]]

In [56]:
movies_fb

director_facebook_likes  actor_3_facebook_likes  actor_1_facebook_likes  \
0                         0.0                   855.0                  1000.0   
1                       563.0                  1000.0                 40000.0   
2                         0.0                   161.0                 11000.0   
3                     22000.0                 23000.0                 27000.0   
4                       131.0                     NaN                   131.0   
...                       ...                     ...                     ...   
4911                      2.0                   318.0                   637.0   
4912                      NaN                   319.0                   841.0   
4913                      0.0                     0.0                     0.0   
4914                      0.0                   489.0                   946.0   
4915                     16.0                    16.0                    86.0   

      cast_total_facebook_likes  actor_2_facebook_likes  movie_facebook_likes  
0                          4834                   936.0                 33000  
1                         48350                  5000.0                     0  
2                         11700                   393.0                 85000  
3                        106759                 23000.0                164000  
4                           143                    12.0                     0  
...                         ...                     ...                   ...  
4911                       2283                   470.0                    84  
4912                       1753                   593.0                 32000  
4913                          0                     0.0                    16  
4914                       2386                   719.0                   660  
4915                        163                    23.0                   456  

[4916 rows x 6 columns]

여기에 모든 수치를 합친 sum 을 계산해 추가해보겠습니다. sum(axis=1)을 응용해봅시다.

assign 함수는 새 컬럼의 이름과 값을 넣어 df 에 합쳐주는 기능을 수행합니다.

In [67]:
movies_fb.sum(axis=1)

0        40625.0
1        94913.0
2       108254.0
3       365759.0
4          417.0
          ...   
4911      3794.0
4912     35506.0
4913        16.0
4914      5200.0
4915       760.0
Length: 4916, dtype: float64

In [58]:
movies_fb = movies_fb.assign(total_likes=movies_fb.sum(axis=1))
movies_fb.head()

director_facebook_likes  actor_3_facebook_likes  actor_1_facebook_likes  \
0                      0.0                   855.0                  1000.0   
1                    563.0                  1000.0                 40000.0   
2                      0.0                   161.0                 11000.0   
3                  22000.0                 23000.0                 27000.0   
4                    131.0                     NaN                   131.0   

   cast_total_facebook_likes  actor_2_facebook_likes  movie_facebook_likes  \
0                       4834                   936.0                 33000   
1                      48350                  5000.0                     0   
2                      11700                   393.0                 85000   
3                     106759                 23000.0                164000   
4                        143                    12.0                     0   

   total_likes  
0      40625.0  
1      94913.0  
2     108254.0  
3     365759.0  
4        417.0

마지막으로 해당 컬럼을 다시 지워보죠. 

지울 때는 두 가지 방법이 있습니다.

.drop 함수를 쓰는 것과 del 을 사용하는 것입니다.

drop 은 다음과 같이 사용합니다.


In [59]:
movies_fb.drop(columns='total_likes').head()

director_facebook_likes  actor_3_facebook_likes  actor_1_facebook_likes  \
0                      0.0                   855.0                  1000.0   
1                    563.0                  1000.0                 40000.0   
2                      0.0                   161.0                 11000.0   
3                  22000.0                 23000.0                 27000.0   
4                    131.0                     NaN                   131.0   

   cast_total_facebook_likes  actor_2_facebook_likes  movie_facebook_likes  
0                       4834                   936.0                 33000  
1                      48350                  5000.0                     0  
2                      11700                   393.0                 85000  
3                     106759                 23000.0                164000  
4                        143                    12.0                     0

del 은 
> del movies_fb['total_likes']

와 같이 사용합니다.

하지만, df 계열의 함수적 접근법이 아니고, df 를 반환하지 않아 drop 을 쓰는 것을 추천합니다.

In [60]:
del movies_fb['total_likes']

# 2. 여러 열 선택하기

여러 열을 한 번에 선택해야 할 때엔 몇 가지 방법을 가지고 접근할 수 있습니다.

- 열의 이름을 넣은 리스트를 활용
- loc 
- iloc
- select_dtypes
- filter

뭐가 많죠?

하나씩 차근차근 알아봅시다.

In [13]:
#필요한 import
import pandas as pd
import numpy as np

In [15]:
# movie data
movies = pd.read_csv('movie.csv')

FileNotFoundError: [Errno 2] No such file or directory: 'movie.csv'

In [70]:
#a[[a1,a2,a3,a4......]]
#a.loc[:,[a1,a2,a3....]]
#a.iloc[:,[a1,a2,a3....]]

KeyError: '컬럼'

## 2.1. 열 이름 직접 사용

내가 필요한 열의 이름을 직접 나열하여 가져오는 방법입니다.

아래의 코드를 같이 볼까요?

In [6]:
movie_people_names = [
    'actor_1_name',
    'actor_2_name',
    'actor_3_name',
    'director_name'
]

먼저 리스트로 내가 가져오고 싶은 컬럼명을 나열합니다. 
그 후 이를 index로 넘겨주면 되겠죠?

In [8]:
movies[movie_people_names].head()

NameError: name 'movies' is not defined

"이름을 활용한다" 라는 것은, loc으로도 가능!하다는 의미가 되겠죠?

동일한 컬럼을 loc 을 이용해 가져와봅시다.

In [ ]:
movies.loc[:, movie_people_names].head()

단일 컬럼을 선택할 때는, 내가 원하는 형태에 따라 DF/S를 결정할 수 있습니다.

예를 들어, actor_1_name 을 선택한다 하면, DF로 원할 때엔 list를, S를 원핼 땐 str 을 넘겨줍니다.

무슨 말인지는 코드를 보며 이해해봅시다. Jupyter 에서는 출력되는 모양이 다를겁니다.

(loc 함수를 사용할 때에도 동일한 원칙이 적용됩니다.)

In [11]:
movies['actor_1_name'].head()

NameError: name 'movies' is not defined

In [81]:
movies[['actor_1_name']]

actor_1_name
0         CCH Pounder
1         Johnny Depp
2     Christoph Waltz
3           Tom Hardy
4         Doug Walker
...               ...
4911      Eric Mabius
4912      Natalie Zea
4913      Eva Boehnke
4914        Alan Ruck
4915      John August

[4916 rows x 1 columns]

## 2.2. 메서드를 이용해 열 선택하기

지난 시간에 rename 했던 것 기억 나나요? 

col_map 을 이용해 딕셔너리 형태로 새로운 컬럼 이름을 정했었죠.

이번에는 딕셔너리가 아닌 함수 형태로 먼저 컬럼의 이름을 정리해보겠습니다.


In [ ]:
movies.loc[:, 'actor_1_name'].value_counts().head()

In [ ]:
pd.Series(movies.columns).head()

In [82]:
def shorten(col):
    return(
        str(col)
        .replace('facebook_likes', 'fb')
        .replace('_for_reviews', '')
    )
movies = movies.rename(columns=shorten) #rename 이름 변경
pd.Series(movies.columns).head()

0            color
1    director_name
2       num_critic
3         duration
4      director_fb
dtype: object

In [ ]:
movies.head()

In [ ]:
movies.dtypes.value_counts()

이제 select_dtypes 메서드를 사용해보겠습니다.

include 에는 어떤 데이터 타입을 가져올 것인지 정할 수 있고, 여러 타입을 선택할 때엔 리스트로 만들어줍니다.

숫자만 선택할 땐 __number__ 를 이용하고, 특정 타입을 제거하고 싶을 땐 exclude 로 데이터타입을 정해줍니다.

In [83]:
movies.select_dtypes(include='int64').head()

num_voted_users  cast_total_fb  movie_fb
0           886204           4834     33000
1           471220          48350         0
2           275868          11700     85000
3          1144337         106759    164000
4                8            143         0

In [84]:
movies.select_dtypes(include=['object', 'int64'])

color      director_name      actor_2_name  \
0     Color      James Cameron  Joel David Moore   
1     Color     Gore Verbinski     Orlando Bloom   
2     Color         Sam Mendes      Rory Kinnear   
3     Color  Christopher Nolan    Christian Bale   
4       NaN        Doug Walker        Rob Walker   
...     ...                ...               ...   
4911  Color        Scott Smith     Daphne Zuniga   
4912  Color                NaN     Valorie Curry   
4913  Color   Benjamin Roberds     Maxwell Moody   
4914  Color        Daniel Hsia     Daniel Henney   
4915  Color           Jon Gunn  Brian Herzlinger   

                               genres     actor_1_name  \
0     Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1            Action|Adventure|Fantasy      Johnny Depp   
2           Action|Adventure|Thriller  Christoph Waltz   
3                     Action|Thriller        Tom Hardy   
4                         Documentary      Doug Walker   
...                               ...              ...   
4911                     Comedy|Drama      Eric Mabius   
4912     Crime|Drama|Mystery|Thriller      Natalie Zea   
4913            Drama|Horror|Thriller      Eva Boehnke   
4914             Comedy|Drama|Romance        Alan Ruck   
4915                      Documentary      John August   

                                     movie_title  num_voted_users  \
0                                         Avatar           886204   
1       Pirates of the Caribbean: At World's End           471220   
2                                        Spectre           275868   
3                          The Dark Knight Rises          1144337   
4     Star Wars: Episode VII - The Force Awakens                8   
...                                          ...              ...   
4911                     Signed Sealed Delivered              629   
4912                               The Following            73839   
4913                        A Plague So Pleasant               38   
4914                            Shanghai Calling             1255   
4915                           My Date with Drew             4285   

      cast_total_fb          actor_3_name  \
0              4834             Wes Studi   
1             48350        Jack Davenport   
2             11700      Stephanie Sigman   
3            106759  Joseph Gordon-Levitt   
4               143                   NaN   
...             ...                   ...   
4911           2283          Crystal Lowe   
4912           1753         Sam Underwood   
4913              0        David Chandler   
4914           2386           Eliza Coupe   
4915            163              Jon Gunn   

                                          plot_keywords  \
0                avatar|future|marine|native|paraplegic   
1     goddess|marriage ceremony|marriage proposal|pi...   
2                   bomb|espionage|sequel|spy|terrorist   
3     deception|imprisonment|lawlessness|police offi...   
4                                                   NaN   
...                                                 ...   
4911             fraud|postal worker|prison|theft|trial   
4912       cult|fbi|hideout|prison escape|serial killer   
4913                                                NaN   
4914                                                NaN   
4915  actress name in title|crush|date|four word tit...   

                                        movie_imdb_link language country  \
0     http://www.imdb.com/title/tt0499549/?ref_=fn_t...  English     USA   
1     http://www.imdb.com/title/tt0449088/?ref_=fn_t...  English     USA   
2     http://www.imdb.com/title/tt2379713/?ref_=fn_t...  English      UK   
3     http://www.imdb.com/title/tt1345836/?ref_=fn_t...  English     USA   
4     http://www.imdb.com/title/tt5289954/?ref_=fn_t...      NaN     NaN   
...                                                 ...      ...     ...   
4911  http://www.imdb.com/title/tt3000844/?ref_=fn_t...  English  Canada 

In [ ]:
movies.select_dtypes(include='number').head() #include빼고 가져오기

In [ ]:
movies.select_dtypes(exclude='number').head()

이제는 filter 함수를 적용해보겠습니다.

다른 방법보다 조금 더 다양하게 활용할 수 있다는 장점이 있습니다.

아래의 예를 볼까요?

In [ ]:
cols = [c for c in movies.columns if 'name' in c]
print('cols :', cols)

In [ ]:
movies.filter(items=cols).head()

filter 함수는 regex 를 사용할 수도 있습니다.

정규표현식에 대해서는 다음 자료를 참고해보세요.

https://www.nextree.co.kr/p4327/

만약 숫자를 가진 컬럼들을 모두 선택하고 싶다면 아래와 같이 작성합니다.

In [ ]:
movies.filter(regex=r'\d')
#정규표현식 공부 한 번만 합시다!

filter 함수는 items, like, 그리고 regex 를 인자로 가질 수 있습니다.

like 는 말 그대로 비슷한 텍스트를 가진 컬럼을 찾아줍니다.

자세한 내용은 책 89페이지를 참고하세요.

In [85]:
movies.filter(like='movie') #like찾을때 사용

movie_title  \
0                                         Avatar   
1       Pirates of the Caribbean: At World's End   
2                                        Spectre   
3                          The Dark Knight Rises   
4     Star Wars: Episode VII - The Force Awakens   
...                                          ...   
4911                     Signed Sealed Delivered   
4912                               The Following   
4913                        A Plague So Pleasant   
4914                            Shanghai Calling   
4915                           My Date with Drew   

                                        movie_imdb_link  movie_fb  
0     http://www.imdb.com/title/tt0499549/?ref_=fn_t...     33000  
1     http://www.imdb.com/title/tt0449088/?ref_=fn_t...         0  
2     http://www.imdb.com/title/tt2379713/?ref_=fn_t...     85000  
3     http://www.imdb.com/title/tt1345836/?ref_=fn_t...    164000  
4     http://www.imdb.com/title/tt5289954/?ref_=fn_t...         0  
...                                                 ...       ...  
4911  http://www.imdb.com/title/tt3000844/?ref_=fn_t...        84  
4912  http://www.imdb.com/title/tt2071645/?ref_=fn_t...     32000  
4913  http://www.imdb.com/title/tt2107644/?ref_=fn_t...        16  
4914  http://www.imdb.com/title/tt2070597/?ref_=fn_t...       660  
4915  http://www.imdb.com/title/tt0378407/?ref_=fn_t...       456  

[4916 rows x 3 columns]

# 2.3. 열 이름 정렬

열의 이름을 정렬해서 DF를 정리하고 싶을 때가 있습니다. 

_(심화)책 90~95쪽에는 의미 있는 열을 위주로 정리하는 것에 대한 내용이 있습니다._

여기에서는 컬럼의 순서만 바꿔서 열을 다뤄보는 것만 해보겠습니다.

위에서 가져온 cols 만 가져와보죠.

In [89]:
movies = pd.read_csv('movie.csv')

In [86]:
cols

['actor_1_facebook_likes',
 'actor_2_facebook_likes',
 'actor_3_facebook_likes',
 'director_facebook_likes']

In [90]:
movies[cols]

actor_1_facebook_likes  actor_2_facebook_likes  actor_3_facebook_likes  \
0                     1000.0                   936.0                   855.0   
1                    40000.0                  5000.0                  1000.0   
2                    11000.0                   393.0                   161.0   
3                    27000.0                 23000.0                 23000.0   
4                      131.0                    12.0                     NaN   
...                      ...                     ...                     ...   
4911                   637.0                   470.0                   318.0   
4912                   841.0                   593.0                   319.0   
4913                     0.0                     0.0                     0.0   
4914                   946.0                   719.0                   489.0   
4915                    86.0                    23.0                    16.0   

      director_facebook_likes  
0                         0.0  
1                       563.0  
2                         0.0  
3                     22000.0  
4                       131.0  
...                       ...  
4911                      2.0  
4912                      NaN  
4913                      0.0  
4914                      0.0  
4915                     16.0  

[4916 rows x 4 columns]

In [91]:
names = movies[['actor_2_name', 'actor_1_name', 'actor_3_name', 'director_name']]
names.head() #중복도가능

actor_2_name     actor_1_name          actor_3_name      director_name
0  Joel David Moore      CCH Pounder             Wes Studi      James Cameron
1     Orlando Bloom      Johnny Depp        Jack Davenport     Gore Verbinski
2      Rory Kinnear  Christoph Waltz      Stephanie Sigman         Sam Mendes
3    Christian Bale        Tom Hardy  Joseph Gordon-Levitt  Christopher Nolan
4        Rob Walker      Doug Walker                   NaN        Doug Walker

눈치 채신 분들도 있겠지만, cols에 등장하는 열의 순서대로 df 가 구성되었음을 알 수 있습니다. 그러니까 컬럼을 부르는 순서를 바꾸면, 열의 순서도 바뀌는 것이죠.

아래와 같이 말이죠.

In [95]:
cols.sort
cols

['actor_1_facebook_likes',
 'actor_2_facebook_likes',
 'actor_3_facebook_likes',
 'director_facebook_likes']

cols 가 list 이기 때문에, sort 메서드를 사용해 정렬했습니다.

names 를 다시 호출해보죠.


In [96]:
names[cols].head()

KeyError: "None of [Index(['actor_1_facebook_likes', 'actor_2_facebook_likes',\n       'actor_3_facebook_likes', 'director_facebook_likes'],\n      dtype='object')] are in the [columns]"

자 열의 순서가 바뀌었죠? 이제 여러분이 원하는 순서대로 열을 맞춰 DF를 구성할 수 있을 겁니다 :)



In [97]:
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [98]:
m_cols = list(movies.columns)
m_cols.sort()
m_cols

['actor_1_facebook_likes',
 'actor_1_name',
 'actor_2_facebook_likes',
 'actor_2_name',
 'actor_3_facebook_likes',
 'actor_3_name',
 'aspect_ratio',
 'budget',
 'cast_total_facebook_likes',
 'color',
 'content_rating',
 'country',
 'director_facebook_likes',
 'director_name',
 'duration',
 'facenumber_in_poster',
 'genres',
 'gross',
 'imdb_score',
 'language',
 'movie_facebook_likes',
 'movie_imdb_link',
 'movie_title',
 'num_critic_for_reviews',
 'num_user_for_reviews',
 'num_voted_users',
 'plot_keywords',
 'title_year']

In [99]:
movies[m_cols].head()

actor_1_facebook_likes     actor_1_name  actor_2_facebook_likes  \
0                  1000.0      CCH Pounder                   936.0   
1                 40000.0      Johnny Depp                  5000.0   
2                 11000.0  Christoph Waltz                   393.0   
3                 27000.0        Tom Hardy                 23000.0   
4                   131.0      Doug Walker                    12.0   

       actor_2_name  actor_3_facebook_likes          actor_3_name  \
0  Joel David Moore                   855.0             Wes Studi   
1     Orlando Bloom                  1000.0        Jack Davenport   
2      Rory Kinnear                   161.0      Stephanie Sigman   
3    Christian Bale                 23000.0  Joseph Gordon-Levitt   
4        Rob Walker                     NaN                   NaN   

   aspect_ratio       budget  cast_total_facebook_likes  color  ...  \
0          1.78  237000000.0                       4834  Color  ...   
1          2.35  300000000.0                      48350  Color  ...   
2          2.35  245000000.0                      11700  Color  ...   
3          2.35  250000000.0                     106759  Color  ...   
4           NaN          NaN                        143    NaN  ...   

  imdb_score language  movie_facebook_likes  \
0        7.9  English                 33000   
1        7.1  English                     0   
2        6.8  English                 85000   
3        8.5  English                164000   
4        7.1      NaN                     0   

                                     movie_imdb_link  \
0  http://www.imdb.com/title/tt0499549/?ref_=fn_t...   
1  http://www.imdb.com/title/tt0449088/?ref_=fn_t...   
2  http://www.imdb.com/title/tt2379713/?ref_=fn_t...   
3  http://www.imdb.com/title/tt1345836/?ref_=fn_t...   
4  http://www.imdb.com/title/tt5289954/?ref_=fn_t...   

                                  movie_title  num_critic_for_reviews  \
0                                      Avatar                   723.0   
1    Pirates of the Caribbean: At World's End                   302.0   
2                                     Spectre                   602.0   
3                       The Dark Knight Rises                   813.0   
4  Star Wars: Episode VII - The Force Awakens                     NaN   

  num_user_for_reviews  num_voted_users  \
0               3054.0           886204   
1               1238.0           471220   
2                994.0           275868   
3               2701.0          1144337   
4                  NaN                8   

                                       plot_keywords title_year  
0             avatar|future|marine|native|paraplegic     2009.0  
1  goddess|marriage ceremony|marriage proposal|pi...     2007.0  
2                bomb|espionage|sequel|spy|terrorist     2015.0  
3  deception|imprisonment|lawlessness|police offi...     2012.0  
4                                                NaN        NaN  

[5 rows x 28 columns]

In [103]:
m_cols[-1] #가장뒤에있는애를가져옴

'title_year'

In [104]:
m_cols[:-1] 

['actor_1_facebook_likes',
 'actor_1_name',
 'actor_2_facebook_likes',
 'actor_2_name',
 'actor_3_facebook_likes',
 'actor_3_name',
 'aspect_ratio',
 'budget',
 'cast_total_facebook_likes',
 'color',
 'content_rating',
 'country',
 'director_facebook_likes',
 'director_name',
 'duration',
 'facenumber_in_poster',
 'genres',
 'gross',
 'imdb_score',
 'language',
 'movie_facebook_likes',
 'movie_imdb_link',
 'movie_title',
 'num_critic_for_reviews',
 'num_user_for_reviews',
 'num_voted_users',
 'plot_keywords']

In [ ]:
# 0~9
# 9, 1~8

In [101]:
movies[[m_cols[-1]] + m_cols[:-1]].head() #가장뒤에있는거가져옴(리스트화)+맨뒤에하나뺴고가져옴

title_year  actor_1_facebook_likes     actor_1_name  \
0      2009.0                  1000.0      CCH Pounder   
1      2007.0                 40000.0      Johnny Depp   
2      2015.0                 11000.0  Christoph Waltz   
3      2012.0                 27000.0        Tom Hardy   
4         NaN                   131.0      Doug Walker   

   actor_2_facebook_likes      actor_2_name  actor_3_facebook_likes  \
0                   936.0  Joel David Moore                   855.0   
1                  5000.0     Orlando Bloom                  1000.0   
2                   393.0      Rory Kinnear                   161.0   
3                 23000.0    Christian Bale                 23000.0   
4                    12.0        Rob Walker                     NaN   

           actor_3_name  aspect_ratio       budget  cast_total_facebook_likes  \
0             Wes Studi          1.78  237000000.0                       4834   
1        Jack Davenport          2.35  300000000.0                      48350   
2      Stephanie Sigman          2.35  245000000.0                      11700   
3  Joseph Gordon-Levitt          2.35  250000000.0                     106759   
4                   NaN           NaN          NaN                        143   

   ...        gross imdb_score language  movie_facebook_likes  \
0  ...  760505847.0        7.9  English                 33000   
1  ...  309404152.0        7.1  English                     0   
2  ...  200074175.0        6.8  English                 85000   
3  ...  448130642.0        8.5  English                164000   
4  ...          NaN        7.1      NaN                     0   

                                     movie_imdb_link  \
0  http://www.imdb.com/title/tt0499549/?ref_=fn_t...   
1  http://www.imdb.com/title/tt0449088/?ref_=fn_t...   
2  http://www.imdb.com/title/tt2379713/?ref_=fn_t...   
3  http://www.imdb.com/title/tt1345836/?ref_=fn_t...   
4  http://www.imdb.com/title/tt5289954/?ref_=fn_t...   

                                  movie_title  num_critic_for_reviews  \
0                                      Avatar                   723.0   
1    Pirates of the Caribbean: At World's End                   302.0   
2                                     Spectre                   602.0   
3                       The Dark Knight Rises                   813.0   
4  Star Wars: Episode VII - The Force Awakens                     NaN   

  num_user_for_reviews  num_voted_users  \
0               3054.0           886204   
1               1238.0           471220   
2                994.0           275868   
3               2701.0          1144337   
4                  NaN                8   

                                       plot_keywords  
0             avatar|future|marine|native|paraplegic  
1  goddess|marriage ceremony|marriage proposal|pi...  
2                bomb|espionage|sequel|spy|terrorist  
3  deception|imprisonment|lawlessness|police offi...  
4                                                NaN  

[5 rows x 28 columns]

In [ ]:
[1] + [2, 3]

In [105]:
#m_cols
# 0~9 -> 3, 0~9 중 3을 빼고 넣기

movies[['color'] + [c for c in m_cols if c != 'color']].head()

color  actor_1_facebook_likes     actor_1_name  actor_2_facebook_likes  \
0  Color                  1000.0      CCH Pounder                   936.0   
1  Color                 40000.0      Johnny Depp                  5000.0   
2  Color                 11000.0  Christoph Waltz                   393.0   
3  Color                 27000.0        Tom Hardy                 23000.0   
4    NaN                   131.0      Doug Walker                    12.0   

       actor_2_name  actor_3_facebook_likes          actor_3_name  \
0  Joel David Moore                   855.0             Wes Studi   
1     Orlando Bloom                  1000.0        Jack Davenport   
2      Rory Kinnear                   161.0      Stephanie Sigman   
3    Christian Bale                 23000.0  Joseph Gordon-Levitt   
4        Rob Walker                     NaN                   NaN   

   aspect_ratio       budget  cast_total_facebook_likes  ... imdb_score  \
0          1.78  237000000.0                       4834  ...        7.9   
1          2.35  300000000.0                      48350  ...        7.1   
2          2.35  245000000.0                      11700  ...        6.8   
3          2.35  250000000.0                     106759  ...        8.5   
4           NaN          NaN                        143  ...        7.1   

  language  movie_facebook_likes  \
0  English                 33000   
1  English                     0   
2  English                 85000   
3  English                164000   
4      NaN                     0   

                                     movie_imdb_link  \
0  http://www.imdb.com/title/tt0499549/?ref_=fn_t...   
1  http://www.imdb.com/title/tt0449088/?ref_=fn_t...   
2  http://www.imdb.com/title/tt2379713/?ref_=fn_t...   
3  http://www.imdb.com/title/tt1345836/?ref_=fn_t...   
4  http://www.imdb.com/title/tt5289954/?ref_=fn_t...   

                                  movie_title  num_critic_for_reviews  \
0                                      Avatar                   723.0   
1    Pirates of the Caribbean: At World's End                   302.0   
2                                     Spectre                   602.0   
3                       The Dark Knight Rises                   813.0   
4  Star Wars: Episode VII - The Force Awakens                     NaN   

  num_user_for_reviews  num_voted_users  \
0               3054.0           886204   
1               1238.0           471220   
2                994.0           275868   
3               2701.0          1144337   
4                  NaN                8   

                                       plot_keywords title_year  
0             avatar|future|marine|native|paraplegic     2009.0  
1  goddess|marriage ceremony|marriage proposal|pi...     2007.0  
2                bomb|espionage|sequel|spy|terrorist     2015.0  
3  deception|imprisonment|lawlessness|police offi...     2012.0  
4                                                NaN        NaN  

[5 rows x 28 columns]

## 2.4. DF 요약

DF 정보를 보여주는 다양한 함수들에 대해 배웠었죠?

여기서는 그에 대해 좀 더 알아보겠습니다.

먼저 shape, size, len, 그리고 ndim 을 볼까요?

In [106]:
movies.head()

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [109]:
# 행, 열 정보를 줍니다.
movies.shape[1]

28

In [110]:
# 행 x 열, 즉 총 데이터의 개수를 세어 줍니다.
movies.size

137648

In [111]:
# row(열) 의 개수를 나타냅니다.
len(movies)

4916

In [112]:
# 몇 차원인지 알려줍니다.(나중에바뀔수있음)
movies.ndim

2

시리즈의 정보를 한 개의 값을 바꿔주던 함수들 기억 나시나요?

count, min, max, mean, median, std 같은 것들이요.

이런 친구들을 집계함수(aggregation function)이라고 합니다.

DF에 이런 함수들을 적용하면, 각 컬럼별로 계산해줍니다.

결과값은 Series 에요.

In [116]:
movies.count().head()

color                      4897
director_name              4814
num_critic_for_reviews     4867
duration                   4901
director_facebook_likes    4814
dtype: int64

In [114]:
movies.mean() #지금나오는 빨간색 : 경고성의미

C:\Users\eyeco\AppData\Local\Temp\ipykernel_30404\1859825645.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  movies.mean()


num_critic_for_reviews       1.379889e+02
duration                     1.070908e+02
director_facebook_likes      6.910145e+02
actor_3_facebook_likes       6.312763e+02
actor_1_facebook_likes       6.494488e+03
gross                        4.764451e+07
num_voted_users              8.264492e+04
cast_total_facebook_likes    9.579816e+03
facenumber_in_poster         1.377320e+00
num_user_for_reviews         2.676688e+02
budget                       3.654749e+07
title_year                   2.002448e+03
actor_2_facebook_likes       1.621924e+03
imdb_score                   6.437429e+00
aspect_ratio                 2.222349e+00
movie_facebook_likes         7.348294e+03
dtype: float64

In [ ]:
selet.dytpyes

describe 는 매우매우매우 중요한 함수입니다. 

DF를 한 번에 정리해주니 참 편리하기도 하고요.

(아래의 .T는 Transpose 의 약자로 행과 열을 맞바꿔주는 기능을 합니다. describe 할 때 요긴해요)

In [120]:
movies.describe().T #스크롤방향을 바꿔줌(위아래->왼쪽오른쪽)

count          mean           std      min  \
num_critic_for_reviews     4867.0  1.379889e+02  1.202394e+02     1.00   
duration                   4901.0  1.070908e+02  2.528602e+01     7.00   
director_facebook_likes    4814.0  6.910145e+02  2.832954e+03     0.00   
actor_3_facebook_likes     4893.0  6.312763e+02  1.625875e+03     0.00   
actor_1_facebook_likes     4909.0  6.494488e+03  1.510699e+04     0.00   
gross                      4054.0  4.764451e+07  6.737255e+07   162.00   
num_voted_users            4916.0  8.264492e+04  1.383222e+05     5.00   
cast_total_facebook_likes  4916.0  9.579816e+03  1.816432e+04     0.00   
facenumber_in_poster       4903.0  1.377320e+00  2.023826e+00     0.00   
num_user_for_reviews       4895.0  2.676688e+02  3.729348e+02     1.00   
budget                     4432.0  3.654749e+07  1.002427e+08   218.00   
title_year                 4810.0  2.002448e+03  1.245398e+01  1916.00   
actor_2_facebook_likes     4903.0  1.621924e+03  4.011300e+03     0.00   
imdb_score                 4916.0  6.437429e+00  1.127802e+00     1.60   
aspect_ratio               4590.0  2.222349e+00  1.402940e+00     1.18   
movie_facebook_likes       4916.0  7.348294e+03  1.920602e+04     0.00   

                                  25%          50%          75%           max  
num_critic_for_reviews          49.00       108.00       191.00  8.130000e+02  
duration                        93.00       103.00       118.00  5.110000e+02  
director_facebook_likes          7.00        48.00       189.75  2.300000e+04  
actor_3_facebook_likes         132.00       366.00       633.00  2.300000e+04  
actor_1_facebook_likes         607.00       982.00     11000.00  6.400000e+05  
gross                      5019656.25  25043962.00  61108412.75  7.605058e+08  
num_voted_users               8361.75     33132.50     93772.75  1.689764e+06  
cast_total_facebook_likes     1394.75      3049.00     13616.75  6.567300e+05  
facenumber_in_poster             0.00         1.00         2.00  4.300000e+01  
num_user_for_reviews            64.00       153.00       320.50  5.060000e+03  
budget                     6000000.00  19850000.00  43000000.00  4.200000e+09  
title_year                    1999.00      2005.00      2011.00  2.016000e+03  
actor_2_facebook_likes         277.00       593.00       912.00  1.370000e+05  
imdb_score                       5.80         6.60         7.20  9.500000e+00  
aspect_ratio                     1.85         2.35         2.35  1.600000e+01  
movie_facebook_likes             0.00       159.00      2000.00  3.490000e+05

filter 에서 사용한 include/exclude 기억나시나요? describe 에서도 쓸 수 있어요.

다음을 볼까요?

In [121]:
movies.describe(include='object').T

count unique  \
color            4897      2   
director_name    4814   2397   
actor_2_name     4903   3030   
genres           4916    914   
actor_1_name     4909   2095   
movie_title      4916   4916   
actor_3_name     4893   3519   
plot_keywords    4764   4756   
movie_imdb_link  4916   4916   
language         4904     47   
country          4911     65   
content_rating   4616     18   

                                                               top  freq  
color                                                        Color  4693  
director_name                                     Steven Spielberg    26  
actor_2_name                                        Morgan Freeman    18  
genres                                                       Drama   233  
actor_1_name                                        Robert De Niro    48  
movie_title                                                 Avatar     1  
actor_3_name                                          Steve Coogan     8  
plot_keywords                                       based on novel     4  
movie_imdb_link  http://www.imdb.com/title/tt0499549/?ref_=fn_t...     1  
language                                                   English  4582  
country                                                        USA  3710  
content_rating                                                   R  2067

## 2.5. DF 메서드 체인

시리즈와 마찬가지로 DF에도 여러 메서드들을 연결해서 적용할 수 있다는 것은 알고 계시죠?

여기서는 이 특성을 이용해 null 값을 다뤄보겠습니다.



In [123]:
movies = pd.read_csv('movie.csv')
def shorten(col):
    return(col.replace('facebook_likes', 'fb').replace('_for_reviews', ''))

movies = movies.rename(columns=shorten)

In [125]:
movies.isna()

color  director_name  num_critic  duration  director_fb  actor_3_fb  \
0     False          False       False     False        False       False   
1     False          False       False     False        False       False   
2     False          False       False     False        False       False   
3     False          False       False     False        False       False   
4      True          False        True      True        False        True   
...     ...            ...         ...       ...          ...         ...   
4911  False          False       False     False        False       False   
4912  False           True       False     False         True       False   
4913  False          False       False     False        False       False   
4914  False          False       False     False        False       False   
4915  False          False       False     False        False       False   

      actor_2_name  actor_1_fb  gross  genres  ...  num_user  language  \
0            False       False  False   False  ...     False     False   
1            False       False  False   False  ...     False     False   
2            False       False  False   False  ...     False     False   
3            False       False  False   False  ...     False     False   
4            False       False   True   False  ...      True      True   
...            ...         ...    ...     ...  ...       ...       ...   
4911         False       False   True   False  ...     False     False   
4912         False       False   True   False  ...     False     False   
4913         False       False   True   False  ...     False     False   
4914         False       False  False   False  ...     False     False   
4915         False       False  False   False  ...     False     False   

      country  content_rating  budget  title_year  actor_2_fb  imdb_score  \
0       False           False   False       False       False       False   
1       False           False   False       False       False       False   
2       False           False   False       False       False       False   
3       False           False   False       False       False       False   
4        True            True    True        True       False       False   
...       ...             ...     ...         ...         ...         ...   
4911    False            True    True       False       False       False   
4912    False           False    True        True       False       False   
4913    False            True   False       False       False       False   
4914    False           False    True       False       False       False   
4915    False           False   False       False       False       False   

      aspect_ratio  movie_fb  
0            False     False  
1            False     False  
2            False     False  
3            False     False  
4             True     False  
...            ...       ...  
4911          True     False  
4912         False     False  
4913          True     False  
4914         False     False  
4915         False     False  

[4916 rows x 28 columns]

여기에 sum 을 적용하면 행별/열별로 결측치를 측정할 수 있다고 했죠?

(axis=0 / 1  기억나세요?)

In [126]:
movies.isna().sum(axis=1)
#pd.DataFrame.sum()

0        0
1        0
2        0
3        0
4       14
        ..
4911     4
4912     5
4913     4
4914     2
4915     0
Length: 4916, dtype: int64

In [127]:
movies.isna().sum(axis=1).sum()

2654

In [128]:
movies.isnull().sum()

color                    19
director_name           102
num_critic               49
duration                 15
director_fb             102
actor_3_fb               23
actor_2_name             13
actor_1_fb                7
gross                   862
genres                    0
actor_1_name              7
movie_title               0
num_voted_users           0
cast_total_fb             0
actor_3_name             23
facenumber_in_poster     13
plot_keywords           152
movie_imdb_link           0
num_user                 21
language                 12
country                   5
content_rating          300
budget                  484
title_year              106
actor_2_fb               13
imdb_score                0
aspect_ratio            326
movie_fb                  0
dtype: int64

여기에 다시 한 번 sum 을 넣으면, 시리즈의 숫자를 모두 더해주겠죠.

그런데, 각 컬럼별 결측치를 정리한 것이니, 이것의 합은 데이터에 존재하는 총 결측치의 수가 됩니다.

In [129]:
movies.isnull().sum().sum()

2654

DF에 결측치가 존재하는지 빠르게 확인하는 방법은, any 를 쓰는 것입니다.

말 그대로, 어떤 것이든 True 가 있으면 True 를 반환하는 함수입니다.

두 번 쓰는 것은 열과 행을 각각 보기 위함이죠.

온전한 데이터라면 isnull 의 결과로 모두 False 가 나올테고요.

In [135]:
all([True, False, True])
#전부다 트루여야 트루

False

TypeError: 'builtin_function_or_method' object is not subscriptable

In [131]:
any([True, False, True])
#하나라도 트루면 트루

True

In [140]:
movies.isnull().any().any() #첫번째any 하나라도 na값이 있니 / 두번째 any 그것들중에 na값이 있니

True

## 2.6. DF 연산

지난 주에 Series 에 scalar 값을 연산에 사용하면, 알아서 복사가 되는 broadcasting 에 대해 다뤘습니다. 그런데 같은 일을 DF에 적용하면 어떻게 될까요?

여기서는 새 데이터인 college.csv 를 이용해보죠.
대학 진학과 인구 통계학적 구성에 대한 자료입니다.

In [142]:
colleges = pd.read_csv('data/college.csv')
colleges.head()

INSTNM        CITY STABBR  HBCU  MENONLY  \
0             Alabama A & M University      Normal     AL   1.0      0.0   
1  University of Alabama at Birmingham  Birmingham     AL   0.0      0.0   
2                   Amridge University  Montgomery     AL   0.0      0.0   
3  University of Alabama in Huntsville  Huntsville     AL   0.0      0.0   
4             Alabama State University  Montgomery     AL   1.0      0.0   

   WOMENONLY  RELAFFIL  SATVRMID  SATMTMID  DISTANCEONLY  ...  UGDS_2MOR  \
0        0.0         0     424.0     420.0           0.0  ...     0.0000   
1        0.0         0     570.0     565.0           0.0  ...     0.0368   
2        0.0         1       NaN       NaN           1.0  ...     0.0000   
3        0.0         0     595.0     590.0           0.0  ...     0.0172   
4        0.0         0     425.0     430.0           0.0  ...     0.0098   

   UGDS_NRA  UGDS_UNKN  PPTUG_EF  CURROPER  PCTPELL  PCTFLOAN  UG25ABV  \
0    0.0059     0.0138    0.0656         1   0.7356    0.8284   0.1049   
1    0.0179     0.0100    0.2607         1   0.3460    0.5214   0.2422   
2    0.0000     0.2715    0.4536         1   0.6801    0.7795   0.8540   
3    0.0332     0.0350    0.2146         1   0.3072    0.4596   0.2640   
4    0.0243     0.0137    0.0892         1   0.7347    0.7554   0.1270   

   MD_EARN_WNE_P10  GRAD_DEBT_MDN_SUPP  
0            30300               33888  
1            39700             21941.5  
2            40100               23370  
3            45500               24097  
4            26600             33118.5  

[5 rows x 27 columns]

In [143]:
colleges.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7535 entries, 0 to 7534
Data columns (total 27 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   INSTNM              7535 non-null   object 
 1   CITY                7535 non-null   object 
 2   STABBR              7535 non-null   object 
 3   HBCU                7164 non-null   float64
 4   MENONLY             7164 non-null   float64
 5   WOMENONLY           7164 non-null   float64
 6   RELAFFIL            7535 non-null   int64  
 7   SATVRMID            1185 non-null   float64
 8   SATMTMID            1196 non-null   float64
 9   DISTANCEONLY        7164 non-null   float64
 10  UGDS                6874 non-null   float64
 11  UGDS_WHITE          6874 non-null   float64
 12  UGDS_BLACK          6874 non-null   float64
 13  UGDS_HISP           6874 non-null   float64
 14  UGDS_ASIAN          6874 non-null   float64
 15  UGDS_AIAN           6874 non-null   float64
 16  UGDS_N

In [144]:
'a' + 'b'

'ab'

In [145]:
colleges + 5

TypeError: can only concatenate str (not "int") to str

생각해보면, 문자열에는 숫자를 더하거나 뺄 수 없는 것이죠.

즉 컬럼별로 계산이 달라야하는데, 이를 반영하지 않은 표현인 것입니다.

이를 고려해서 다시 시도해보겠습니다.

여기서는 UGDS_로 시작하는 컬럼만 추출해서 써보죠.

In [159]:
colleges = pd.read_csv('data/college.csv', index_col='INSTNM')
# index_col 에서는 어떤 열을 인덱스로 사용할지 지정할 수 있습니다.
# INSTNM 은 학교 이름입니다.

college_ugds = colleges.filter(like='UGDS_')
college_ugds

UGDS_WHITE  UGDS_BLACK  \
INSTNM                                                                       
Alabama A & M University                                0.0333      0.9353   
University of Alabama at Birmingham                     0.5922      0.2600   
Amridge University                                      0.2990      0.4192   
University of Alabama in Huntsville                     0.6988      0.1255   
Alabama State University                                0.0158      0.9208   
...                                                        ...         ...   
SAE Institute of Technology  San Francisco                 NaN         NaN   
Rasmussen College - Overland Park                          NaN         NaN   
National Personal Training Institute of Cleveland          NaN         NaN   
Bay Area Medical Academy - San Jose Satellite L...         NaN         NaN   
Excel Learning Center-San Antonio South                    NaN         NaN   

                                                    UGDS_HISP  UGDS_ASIAN  \
INSTNM                                                                      
Alabama A & M University                               0.0055      0.0019   
University of Alabama at Birmingham                    0.0283      0.0518   
Amridge University                                     0.0069      0.0034   
University of Alabama in Huntsville                    0.0382      0.0376   
Alabama State University                               0.0121      0.0019   
...                                                       ...         ...   
SAE Institute of Technology  San Francisco                NaN         NaN   
Rasmussen College - Overland Park                         NaN         NaN   
National Personal Training Institute of Cleveland         NaN         NaN   
Bay Area Medical Academy - San Jose Satellite L...        NaN         NaN   
Excel Learning Center-San Antonio South                   NaN         NaN   

                                                    UGDS_AIAN  UGDS_NHPI  \
INSTNM                                                                     
Alabama A & M University                               0.0024     0.0019   
University of Alabama at Birmingham                    0.0022     0.0007   
Amridge University                                     0.0000     0.0000   
University of Alabama in Huntsville                    0.0143     0.0002   
Alabama State University                               0.0010     0.0006   
...                                                       ...        ...   
SAE Institute of Technology  San Francisco                NaN        NaN   
Rasmussen College - Overland Park                         NaN        NaN   
National Personal Training Institute of Cleveland         NaN        NaN   
Bay Area Medical Academy - San Jose Satellite L...        NaN        NaN   
Excel Learning Center-San Antonio South                   NaN        NaN   

                                                    UGDS_2MOR  UGDS_NRA  \
INSTNM                                                                    
Alabama A & M University                               0.0000    0.0059   
University of Alabama at Birmingham                    0.0368    0.0179   
Amridge University                                     0.0000    0.0000   
University of Alabama in Huntsville                    0.0172    0.0332   
Alabama State University                               0.0098    0.0243   
...                                                       ...       ...   
SAE Institute of Technology  San Francisco                NaN       NaN   
Rasmussen College - Overland Park                         NaN       NaN   
National Personal Training Institute of Cleveland         NaN       NaN   
Bay Area Medical Academy - San Jose Satellite L...        NaN       NaN   
Excel Learning Center-San Antonio South                   NaN       NaN   

                                                    UGDS_UNKN  
INSTNM

모든 컬럼이 수치형인것으로 보이긴 합니다만, 확인을 해봐야겠죠?

info를 써봅시다.

In [147]:
college_ugds.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7535 entries, Alabama A & M University to Excel Learning Center-San Antonio South
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   UGDS_WHITE  6874 non-null   float64
 1   UGDS_BLACK  6874 non-null   float64
 2   UGDS_HISP   6874 non-null   float64
 3   UGDS_ASIAN  6874 non-null   float64
 4   UGDS_AIAN   6874 non-null   float64
 5   UGDS_NHPI   6874 non-null   float64
 6   UGDS_2MOR   6874 non-null   float64
 7   UGDS_NRA    6874 non-null   float64
 8   UGDS_UNKN   6874 non-null   float64
dtypes: float64(9)
memory usage: 588.7+ KB


In [148]:
college_ugds + 1

UGDS_WHITE  UGDS_BLACK  \
INSTNM                                                                       
Alabama A & M University                                1.0333      1.9353   
University of Alabama at Birmingham                     1.5922      1.2600   
Amridge University                                      1.2990      1.4192   
University of Alabama in Huntsville                     1.6988      1.1255   
Alabama State University                                1.0158      1.9208   
...                                                        ...         ...   
SAE Institute of Technology  San Francisco                 NaN         NaN   
Rasmussen College - Overland Park                          NaN         NaN   
National Personal Training Institute of Cleveland          NaN         NaN   
Bay Area Medical Academy - San Jose Satellite L...         NaN         NaN   
Excel Learning Center-San Antonio South                    NaN         NaN   

                                                    UGDS_HISP  UGDS_ASIAN  \
INSTNM                                                                      
Alabama A & M University                               1.0055      1.0019   
University of Alabama at Birmingham                    1.0283      1.0518   
Amridge University                                     1.0069      1.0034   
University of Alabama in Huntsville                    1.0382      1.0376   
Alabama State University                               1.0121      1.0019   
...                                                       ...         ...   
SAE Institute of Technology  San Francisco                NaN         NaN   
Rasmussen College - Overland Park                         NaN         NaN   
National Personal Training Institute of Cleveland         NaN         NaN   
Bay Area Medical Academy - San Jose Satellite L...        NaN         NaN   
Excel Learning Center-San Antonio South                   NaN         NaN   

                                                    UGDS_AIAN  UGDS_NHPI  \
INSTNM                                                                     
Alabama A & M University                               1.0024     1.0019   
University of Alabama at Birmingham                    1.0022     1.0007   
Amridge University                                     1.0000     1.0000   
University of Alabama in Huntsville                    1.0143     1.0002   
Alabama State University                               1.0010     1.0006   
...                                                       ...        ...   
SAE Institute of Technology  San Francisco                NaN        NaN   
Rasmussen College - Overland Park                         NaN        NaN   
National Personal Training Institute of Cleveland         NaN        NaN   
Bay Area Medical Academy - San Jose Satellite L...        NaN        NaN   
Excel Learning Center-San Antonio South                   NaN        NaN   

                                                    UGDS_2MOR  UGDS_NRA  \
INSTNM                                                                    
Alabama A & M University                               1.0000    1.0059   
University of Alabama at Birmingham                    1.0368    1.0179   
Amridge University                                     1.0000    1.0000   
University of Alabama in Huntsville                    1.0172    1.0332   
Alabama State University                               1.0098    1.0243   
...                                                       ...       ...   
SAE Institute of Technology  San Francisco                NaN       NaN   
Rasmussen College - Overland Park                         NaN       NaN   
National Personal Training Institute of Cleveland         NaN       NaN   
Bay Area Medical Academy - San Jose Satellite L...        NaN       NaN   
Excel Learning Center-San Antonio South                   NaN       NaN   

                                                    UGDS_UNKN  
INSTNM

모두 float 이네요. 그러니까 이제! 계산에 적용해볼 수 있을 겁니다.

여기선 반올림에 대해 다뤄볼겁니다.

pandas 는 숫자가 정 가운데, 그러니까 0.5에 있으면 가까운 짝수 쪽으로 반올림을 시행합니다.

무슨 말이냐면, 아래의 코드를...

In [149]:
round(0.5) #가까운 짝수로 붙여줌

0

만약에 0.5 를 올리고 싶다면, 이보다 약간 큰 수를 만들어주면 됩니다. 이렇게요.

In [150]:
round(0.5 + 0.000001)

1

이를 ugds 데이터에 적용해보겠습니다.

Northwest-Shoals Community College 라는 학교에 적용을 해볼까요?

filter 로 찾아보면 다음과 같습니다.

In [153]:
college_ugds.filter(like='Northwest-Shoals', axis=0).T

INSTNM      Northwest-Shoals Community College
UGDS_WHITE                              0.7912
UGDS_BLACK                              0.1250
UGDS_HISP                               0.0339
UGDS_ASIAN                              0.0036
UGDS_AIAN                               0.0088
UGDS_NHPI                               0.0006
UGDS_2MOR                               0.0012
UGDS_NRA                                0.0033
UGDS_UNKN                               0.0324

In [154]:
#여기에 반올림을 적용해봅시다.
#UGDS_BLACK 을 잘 보세요

college_ugds.filter(like='Northwest-Shoals', axis=0).T.round(2)


INSTNM      Northwest-Shoals Community College
UGDS_WHITE                                0.79
UGDS_BLACK                                0.12
UGDS_HISP                                 0.03
UGDS_ASIAN                                0.00
UGDS_AIAN                                 0.01
UGDS_NHPI                                 0.00
UGDS_2MOR                                 0.00
UGDS_NRA                                  0.00
UGDS_UNKN                                 0.03

In [155]:
# 이를 막기 위해 작은 수를 더해줍시다.
# add함수를 쓰면 chain 을 깨지 않고 만들 수 있죠!

college_ugds.filter(like='Northwest-Shoals', axis=0).T.add(.00001).round(2)

INSTNM      Northwest-Shoals Community College
UGDS_WHITE                                0.79
UGDS_BLACK                                0.13
UGDS_HISP                                 0.03
UGDS_ASIAN                                0.00
UGDS_AIAN                                 0.01
UGDS_NHPI                                 0.00
UGDS_2MOR                                 0.00
UGDS_NRA                                  0.00
UGDS_UNKN                                 0.03

In [156]:
# 이제 전체 DF에 동일한 작업을 해줍시다.
# 형태가 같으므로, 바로 덧셈이 됩니다!

(college_ugds + 0.00001).round(2)

# chain 식으로 표현한다면 아래와 같이.
#college_ugds.add(0.0001).round(2)

UGDS_WHITE  UGDS_BLACK  \
INSTNM                                                                       
Alabama A & M University                                  0.03        0.94   
University of Alabama at Birmingham                       0.59        0.26   
Amridge University                                        0.30        0.42   
University of Alabama in Huntsville                       0.70        0.13   
Alabama State University                                  0.02        0.92   
...                                                        ...         ...   
SAE Institute of Technology  San Francisco                 NaN         NaN   
Rasmussen College - Overland Park                          NaN         NaN   
National Personal Training Institute of Cleveland          NaN         NaN   
Bay Area Medical Academy - San Jose Satellite L...         NaN         NaN   
Excel Learning Center-San Antonio South                    NaN         NaN   

                                                    UGDS_HISP  UGDS_ASIAN  \
INSTNM                                                                      
Alabama A & M University                                 0.01        0.00   
University of Alabama at Birmingham                      0.03        0.05   
Amridge University                                       0.01        0.00   
University of Alabama in Huntsville                      0.04        0.04   
Alabama State University                                 0.01        0.00   
...                                                       ...         ...   
SAE Institute of Technology  San Francisco                NaN         NaN   
Rasmussen College - Overland Park                         NaN         NaN   
National Personal Training Institute of Cleveland         NaN         NaN   
Bay Area Medical Academy - San Jose Satellite L...        NaN         NaN   
Excel Learning Center-San Antonio South                   NaN         NaN   

                                                    UGDS_AIAN  UGDS_NHPI  \
INSTNM                                                                     
Alabama A & M University                                 0.00        0.0   
University of Alabama at Birmingham                      0.00        0.0   
Amridge University                                       0.00        0.0   
University of Alabama in Huntsville                      0.01        0.0   
Alabama State University                                 0.00        0.0   
...                                                       ...        ...   
SAE Institute of Technology  San Francisco                NaN        NaN   
Rasmussen College - Overland Park                         NaN        NaN   
National Personal Training Institute of Cleveland         NaN        NaN   
Bay Area Medical Academy - San Jose Satellite L...        NaN        NaN   
Excel Learning Center-San Antonio South                   NaN        NaN   

                                                    UGDS_2MOR  UGDS_NRA  \
INSTNM                                                                    
Alabama A & M University                                 0.00      0.01   
University of Alabama at Birmingham                      0.04      0.02   
Amridge University                                       0.00      0.00   
University of Alabama in Huntsville                      0.02      0.03   
Alabama State University                                 0.01      0.02   
...                                                       ...       ...   
SAE Institute of Technology  San Francisco                NaN       NaN   
Rasmussen College - Overland Park                         NaN       NaN   
National Personal Training Institute of Cleveland         NaN       NaN   
Bay Area Medical Academy - San Jose Satellite L...        NaN       NaN   
Excel Learning Center-San Antonio South                   NaN       NaN   

                                                    UGDS_UNKN  
INSTNM

## 1.7. 끝나기 전의 팁 NA는 비교불가

NA 값은 NA 끼리도 비교가 불가능합니다.

아래의 출력을 보죠.

In [157]:
print(np.nan != np.nan)
print(np.nan > 1)
print(np.nan < 1)
print(np.nan == 1)

True
False
False
False


오직 같지 않다! 만 참이 나오고 나머지가 다 거짓이 되죠.

그러니까, S나 DF를 비교할 때도 null 값을 처리하지 않으면 문제가 됩니다.

아래와 같이 말이죠.

In [ ]:
import pandas as pd
a = pd.Series([1, 2, 3])
b = pd.Series([1, 2, 3])
c = pd.Series([1, np.nan, 3])
d = pd.Series([1, np.nan, 3])

In [ ]:
#무사통과하는 a == b
print(a == b)
print()
# 1번 인덱스를 잘 보세요.
print(c == d)

그래서 null 값이 섞여 있는 객체를 비교할 땐, __equals__ 메서드를 사용해야만 합니다.

In [ ]:
# equals 는 두 객체가 동일한 요소들을 갖고 있는지 확인하는 메서드입니다.
print(a.equals(b))
print(c.equals(d))

# 오늘의 과제

1. movies df 에서 'movie' 라는 단어가 들어간 컬럼만 선택하고자 합니다. 어떻게 하면 될까요?
for ~ if 를 사용하는 것과, filter 메서드를 사용하는 것 두 가지 모두 고려하세요.

2. movies df 에서 가장 na 값이 많은 컬럼을 찾고자 합니다.
아래의 힌트를 보고, 어떻게 해야할지 코드를 작성해보세요.

> pd.Series([5, 2, 3, 1, 7]).sort_values(ascending=False)

3. sort_values에서 ascending=True 로 바꾸면 어떤 일이 일어나나요?





In [ ]:
pd.Series([5, 2, 3, 1, 7]).sort_values(ascending=False)

In [ ]:
filter(like='movie')

In [ ]:
movie.isna().sum()